_drafts/c-struct-padding-initialization.md

```diff
@@ -0,0 +1,460 @@
+
+❯ gcc --version
+gcc (Ubuntu 11.2.0-7ubuntu2) 11.2.0
+
+# specific package versions:
+❯ apt list clang gcc
+Listing... Done
+clang/impish,now 1:13.0-53~exp1 amd64 [installed]
+gcc/impish,now 4:11.2.0-1ubuntu1 amd64 [installed]
+```
+
+Padding values under each strategy, optimization level, and compiler (_warning
+boring tables below!_):
+
+### Strategy 1, `memset`
+
+| Strategy    | Optimization Level | Clang 13 | GCC 11 |
+| ----------- | ------------------ | -------- | ------ |
+| 1, `memset` | `0`                | zero     | zero   |
+| 1, `memset` | `1`                | zero     | zero   |
+| 1, `memset` | `2`                | zero     | zero   |
+| 1, `memset` | `3`                | zero     | zero   |
+| 1, `memset` | `s`                | zero     | zero   |
+
+### Strategy 2, explicitly setting each struct member
+
+| Strategy    | Optimization Level | Clang 13  | GCC 11    |
+| ----------- | ------------------ | --------- | --------- |
+| 2, explicit | `0`                | zero      | **unset** |
+| 2, explicit | `1`                | **unset** | **unset** |
+| 2, explicit | `2`                | zero      | **unset** |
+| 2, explicit | `3`                | zero      | **unset** |
+| 2, explicit | `s`                | **unset** | **unset** |
+
+### Strategy 3, `{ 0 }`
+
+| Strategy   | Optimization Level | Clang 13  | GCC 11 |
+| ---------- | ------------------ | --------- | ------ |
+| 3, `{ 0 }` | `0`                | zero      | zero   |
+| 3, `{ 0 }` | `1`                | **unset** | zero   |
+| 3, `{ 0 }` | `2`                | zero      | zero   |
+| 3, `{ 0 }` | `3`                | zero      | zero   |
+| 3, `{ 0 }` | `s`                | zero      | zero   |
+
+### Strategy 4, `{ }` GCC extension
+
+| Strategy | Optimization Level | Clang 13  | GCC 11 |
+| -------- | ------------------ | --------- | ------ |
+| 4, `{ }` | `0`                | zero      | zero   |
+| 4, `{ }` | `1`                | **unset** | zero   |
+| 4, `{ }` | `2`                | zero      | zero   |
+| 4, `{ }` | `3`                | zero      | zero   |
+| 4, `{ }` | `s`                | **unset** | zero   |
+
+The main point is that it's not particulary consistent across compilers and
+optimization levels 😱!
+
+> You can find the example application used to generate the above data here on
+> [Github](https://github.com/memfault/interrupt/tree/master/example/c-struct-padding-initialization/).
+
+> I've also uploaded it to the wonderful Compiler Explorer if you want to take a
+> look and quickly play around: <https://godbolt.org/z/b985G4ejT>
+
+## Best Practice?
+
+It's tricky to recommend a one-size-fits-all option here, because different
+software will have different constraints. However, some general purpose advice
+follows.
+
+### Avoid relying on structure layout
+
+The simplest option to avoid padding issues is to avoid the padding fields
+altogether:
+
+> Access data directly via each member, do not alias structures or use `memcmp`
```

Comment:
Maybe note it is okay to use memcmp if struct is packed though

Reply:
added a callout for this